core/src/main/scala/org/apache/spark/deploy/history/HistoryAppStatusStore.scala

```diff
@@ -0,0 +1,123 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.history
+
+import scala.util.matching.Regex
+
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config.History.{APPLY_CUSTOM_EXECUTOR_LOG_URL_TO_INCOMPLETE_APP, CUSTOM_EXECUTOR_LOG_URL}
+import org.apache.spark.status.{AppStatusListener, AppStatusStore}
+import org.apache.spark.status.api.v1
+import org.apache.spark.status.api.v1.ExecutorSummary
+import org.apache.spark.util.kvstore.KVStore
+
+private[spark] class HistoryAppStatusStore(
+    conf: SparkConf,
+    store: KVStore,
+    listener: Option[AppStatusListener] = None)
+  extends AppStatusStore(store, listener) with Logging {
+
+  import HistoryAppStatusStore._
+
+  private val logUrlPattern: Option[String] = conf.get(CUSTOM_EXECUTOR_LOG_URL)
+  private val applyReplaceLogUrlToIncompleteApp: Boolean =
+    conf.get(APPLY_CUSTOM_EXECUTOR_LOG_URL_TO_INCOMPLETE_APP)
+
+  override def executorList(activeOnly: Boolean): Seq[v1.ExecutorSummary] = {
+    val execList = super.executorList(activeOnly)
+    logUrlPattern match {
+      case Some(pattern) =>
+        if (applyReplaceLogUrlToIncompleteApp || isApplicationCompleted) {
+          execList.map(replaceLogUrls(_, pattern))
+        } else {
+          execList
+        }
+
+      case None => execList
+    }
+  }
+
+  override def executorSummary(executorId: String): v1.ExecutorSummary = {
+    val execSummary = super.executorSummary(executorId)
+    logUrlPattern match {
+      case Some(pattern) =>
+        if (applyReplaceLogUrlToIncompleteApp || isApplicationCompleted) {
+          replaceLogUrls(execSummary, pattern)
+        } else {
+          execSummary
+        }
+
+      case None => execSummary
+    }
+  }
+
+  private def isApplicationCompleted: Boolean = {
+    val appInfo = super.applicationInfo()
+    appInfo.attempts.nonEmpty && appInfo.attempts.head.completed
+  }
+
+  private def replaceLogUrls(exec: ExecutorSummary, urlPattern: String): ExecutorSummary = {
+    val attributes = exec.attributes
+
+    val allPatterns = CUSTOM_URL_PATTERN_REGEX.findAllMatchIn(urlPattern).map(_.group(1)).toSet
```

Comment:
This needs at least a comment about what's going on with `FILE_NAME` and `LOG_FILES`. They're special-cased in a bunch of places but I didn't really find an explanation for that.

I'm also wondering whether you could us `replaceAllIn` here to simplify this code...

Reply:
I'll add comment regarding relation between FILE_NAME pattern and LOG_FILES attribute here.

I'm not sure we can simplify the code block with `replaceAllIn`, because `findAllMatchIn` is intentionally used to find all patterns regardless of availability in attributes and enable fail-back to original log URLs. If we just simply use `replaceAllIn` to only replace matched pattern, final URL will become odd and broken one if there's missing attribute.